### PR TITLE
Occult 2

### LINKED ISSUE #9: OCCULT-2 algorithm
The original idea comes [from here](https://arxiv.org/abs/1307.5046).

A [tutorial](http://www.lmsal.com/~aschwand/software/tracing/tracing_tutorial1.html) and the [latest code](https://hesperia.gsfc.nasa.gov/ssw/packages/mjastereo/idl/looptracing_auto4.pro)

We need to figure out which of those IDL codes is the true method or what the difference is between them.

We think that `tracing_auto` calls `tracing_direction` and `tracing_step` while `tracing_auto2` is a modified version of `tracing_auto` which uses both a high pass and low pass filter. However this would need to be checked. 

Base:
- [ ] Working and well implemented algorithm. 
- [ ] Numerically compared to known working output (as a test).

Extra:
- [ ] Optimized as much as possible for memory and CPU time.
- [ ] 100% test coverage
- [ ] Documentation and a worked example.

### PR DESCRIPTION
**Pull request by vatch123**
### Description

The python version of the occult2.

TODO:
- [x] Review of PR
- [x] ~100% test line coverage 
- [x] Code formatted/linted.
- [x] Example and documentation is detailed.

Fixes #9

### CODE REVIEW

File: sunkit_image/enhance.py [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+    """
+
+    # 1. Supress the background
+    image = background_supression(smap.data, zmin, qmed)
+
+    # 2. Bandpass filter
+    image = bandpass_filter(image, nsm1, nsm2)
+
+    x, y = np.meshgrid(*[np.arange(v.value) for v in smap.dimensions]) * u.pix
+    coords = smap.pixel_to_world(x, y).transform_to(frames.Helioprojective)
```
Comment (wafels) [inline sunkit_image/enhance.py]: Can use the `all_coordinates_from_map` in `sunpy.map.maputils` SunPy 1.0 to get coords

File: sunkit_image/enhance.py [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+
+
+    # Creating the three starting arrays
+    n_s = 100
```
Comment (wafels) [inline sunkit_image/enhance.py]: It would be good if you could provide references to the equations, or paragraphs in the paper for what these variables are.  Also, variable names can be much more descriptive.  There is no need to stick to what is in the IDL code or in the paper.  It would be much better to have descriptive names so that later users can understand what is going on.

File: sunkit_image/enhance.py [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+    n_r = 30
+    r_m = (rmin / (-1 + np.arange(n_r) * (2 / n_r - 1))).reshape((-1, 1))
+
+    loops = []
+    ngaps = 3               # Number of empty pixels to denote the end of loop
+
+    # Loops tracing begin
+    for i in range(num_loop):
+
+        z_0 = image.max()
```
Comment (wafels) [inline sunkit_image/enhance.py]: I'd like to see an explanation for every line here so we know what is going on and why.

File: sunkit_image/enhance.py [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+                y_k = y_0
+
+
+        # Zero out the loop pixels assuming values are stored in pixels format
+        for points in loop:
+            image[points[0], points[1]] = 0
+
+        loops.append(loop)
+
+    # Return a map with loops marked. What intensity values to be used for marking loops?
```
Comment (wafels) [inline sunkit_image/enhance.py]: I think the return should be a list of loops, where each loop is a SkyCoord object.  Drawing on the map can be done later.

File: sunkit_image/enhance.py (hunk 2 above)
Comment (vatch123) [reply sunkit_image/enhance.py]: Okay I have tried to use the exact variable names as mentioned in the paper. Will see to add more verbosity.

File: sunkit_image/enhance.py (hunk 3 above)
Comment (vatch123) [reply sunkit_image/enhance.py]: Okay I will add those in comments when I will write the complete code when everything is clear.

File: sunkit_image/enhance.py (hunk 4 above)
Comment (vatch123) [reply sunkit_image/enhance.py]: Okay

File: sunkit_image/enhance.py [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+
+    Returns
+    -------
+    loops : `list`
+        A list of all loop where each element is a `astropy.coordinates.SkyCoord` object
+
+    References
+    ----------
+    * Markus J. Aschwanden, Bart De Pontieu, Eugene A. Katrukha.
+      Optimization of Curvi-Linear Tracing Applied to Solar Physics and Biophysics.
```
Comment (wafels) [inline sunkit_image/enhance.py]: should have the full reference here to the journal and also the arxiv.org link

File: changelog/31.feature.rst [hunk 6]
```diff
@@ -0,0 +1 @@
+Added a new function (`sunkit_image.enhance.occult2`) to automatically trace out loops in a solar image.
```
Comment (nabobalis) [inline changelog/31.feature.rst]:
```suggestion
Added a new function (`sunkit_image.enhance.occult2`) to automatically trace out loops/curved structures in an image.
```

**Comment by nabobalis** [comment]
I think the occult code should be in a new file called trace or something better.

File: docs/code_ref/noise.rst [hunk 7]
```diff
@@ -0,0 +1,8 @@
+*****
+noise
+*****
+
+This package implements functions for noise level estimation.
+
+.. automodapi:: sunkit_image.noise
+   :no-inheritance-diagram:
```
Comment (nabobalis) [inline docs/code_ref/noise.rst]: Why is this here?

File: sunkit_image/check.py [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+#     ima[4, 7] = 1.0
+#     ima[5, 8] = 1.0
+#     ima[4, 9] = 1.0
+#     ima[3, 10] = 1.0
+#     return ima
+
+# image = test_image()
+image = astropy.io.fits.getdata("http://www.lmsal.com/~aschwand/software/tracing/TRACE_19980519.fits", ignore_missing_end=True)
+
+loops = trace.occult2(image, nsm1=3, rmin=30, lmin=25, nstruc=1000, nloop=1000, ngap=0, qthresh1=0.0, qthresh2=3.0, file=True)
```
Comment (nabobalis) [inline sunkit_image/check.py]: Time to remove this file.

File: examples/tracing_loops.py [hunk 9]
```diff
@@ -0,0 +1,63 @@
+"""
+=====================
+Coronal Loops Tracing
+=====================
+
+This example applies traces out the coronal loops in a FITS image file
```
Comment (nabobalis) [inline examples/tracing_loops.py]:
```suggestion
This example applies traces out the coronal loops in a FITS file
```

File: examples/tracing_loops.py [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+using `sunkit_image.trace.occult2`. In this example we will use the settings
```
Comment (nabobalis) [inline examples/tracing_loops.py]:
```suggestion
using `~sunkit_image.trace.occult2`. In this example we will use the settings
```

File: examples/tracing_loops.py [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+and the data from Markus J. Aschwanden tutorial on his IDL implemetation of
```
Comment (nabobalis) [inline examples/tracing_loops.py]:
```suggestion
and the data from Markus Aschwanden's tutorial on his IDL implementation of
```

File: examples/tracing_loops.py [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+the ``OCCULT2`` algorithm, which can be found `here <http://www.lmsal.com/~aschwand/software/tracing/tracing_tutorial1.html>__.
```
Comment (nabobalis) [inline examples/tracing_loops.py]:
```suggestion
the ``OCCULT2`` algorithm, which can be found `here <http://www.lmsal.com/~aschwand/software/tracing/tracing_tutorial1.html>__.
```

File: examples/tracing_loops.py [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+The aim of this example is to demonstrate that `sunkit_image.trace.occult2` provides similar
```
Comment (nabobalis) [inline examples/tracing_loops.py]:
```suggestion
The aim of this example is to demonstrate that `~sunkit_image.trace.occult2` provides similar
```

File: examples/tracing_loops.py (hunk 12 above)
Comment (nabobalis) [inline examples/tracing_loops.py]:
```suggestion
the ``OCCULT2`` algorithm, which can be found `here <http://www.lmsal.com/~aschwand/software/tracing/tracing_tutorial1.html>`__.
```

File: examples/tracing_loops.py [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+# sphinx_gallery_thumbnail_number = 2
+
+import matplotlib.pyplot as plt
+import astropy
+import numpy as np
+
+import sunkit_image.trace as trace
+
+###########################################################################
+# We will be using `astropy.io.fits.getdata` to read a FITS file from the tutorial website
```
Comment (nabobalis) [inline examples/tracing_loops.py]:
```suggestion
# We will be using `astropy.io.fits.getdata` to read the FITS file from the tutorial website.
```

File: examples/tracing_loops.py [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+image = astropy.io.fits.getdata("http://www.lmsal.com/~aschwand/software/tracing/TRACE_19980519.fits", ignore_missing_end=True)
+
+# The original image is plotted to showcase the difference. Matplotlib plots are inverted by default so
```
Comment (nabobalis) [inline examples/tracing_loops.py]:
```suggestion
# The original image is plotted to showcase the difference.
```
What difference?

File: examples/tracing_loops.py [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+# to fix it we will be using ``origin`` as ``lower``.
```
Comment (nabobalis) [inline examples/tracing_loops.py]:
```suggestion
```

File: examples/tracing_loops.py [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+plt.imshow(image, cmap="Spectral", origin="lower")
```
Comment (nabobalis) [inline examples/tracing_loops.py]:
```suggestion
plt.imshow(image, cmap="Spectral", origin="lower")
```

Is that the correct colour map? @wafels, do we have the colormap in SunPy?

File: examples/tracing_loops.py [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+# as in the IDL tutorial.
+# * The lowpass filter boxcar filter size ``nsm1`` is taken to be 3.
+# * The minimum radius of curvature at any point in the loop ``rmin`` is 30 pixels.
+# * The length of the smallest loop to be detected ``lmin`` is 25 pixels.
+# * The maximum number of structures to be examined ``nstruc`` is 1000.
+# * The maximum number of loops to be detected ``nloop`` is 1000.
+# * The number of extra points in the loop below noise level to terminate a loop tracing ``ngap`` is 0.
+# * The base flux and median flux ratio ``qthresh1`` is 0.0
+# * The noise threshold in the image with repect to median flux ``qthresh2`` is 3.0 
+loops = trace.occult2(image, nsm1=3, rmin=30, lmin=25, nstruc=1000, nloop=1000, ngap=0, qthresh1=0.0, qthresh2=3.0, file=True)
```
Comment (nabobalis) [inline examples/tracing_loops.py]:
```suggestion
loops = trace.occult2(image, nsm1=3, rmin=30, lmin=25, nstruc=1000, nloop=1000, 
                                  ngap=0, qthresh1=0.0, qthresh2=3.0, file=True)
```

File: examples/tracing_loops.py [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+
+###############################################################################
+# The `sunkit_image.trace.occult2` returns a list of all loop where each element
```
Comment (nabobalis) [inline examples/tracing_loops.py]:
```suggestion
# `~sunkit_image.trace.occult2` returns a list of all loop where each element
```

File: examples/tracing_loops.py [hunk 20]
```diff
@@ -0,0 +1,63 @@
...
+for loop in loops:
+
+    # We collect all the ``x`` and ``y`` coordinates in seperate lists for plotting.
+    x = []
+    y = []
+    for points in loop:
+        x.append(points[0])
+        y.append(points[1])
+    
+    plt.plot(x, y, 'b')
```
Comment (nabobalis) [inline examples/tracing_loops.py]:
```suggestion
    plt.plot(x, y, 'b')
    
```

There has to be a better way to do this.

File: examples/tracing_loops.py [hunk 21]
```diff
@@ -0,0 +1,63 @@
...
+#################################################################################
+# We also plot the original image to overlay the loops over it.
+plt.imshow(image, cmap="Spectral", origin="lower")
```
Comment (nabobalis) [inline examples/tracing_loops.py]:
```suggestion

plt.imshow(image, cmap="Spectral", origin="lower")
```

File: examples/tracing_loops.py [hunk 22]
```diff
@@ -0,0 +1,63 @@
...
+
+# All the plots are plotted at the end
```
Comment (nabobalis) [inline examples/tracing_loops.py]:
```suggestion
```

File: sunkit_image/trace.py [hunk 23]
```diff
@@ -0,0 +1,195 @@
+"""
+This module contains functions that will enhance the trace out structures in an
+image.
```
Comment (nabobalis) [inline sunkit_image/trace.py]:
```suggestion
```

File: sunkit_image/trace.py [hunk 24]
```diff
@@ -0,0 +1,195 @@
+"""
+This module contains functions that will enhance the trace out structures in an
```
Comment (nabobalis) [inline sunkit_image/trace.py]:
```suggestion
This module contains functions that will  the trace out structures in an image.
```

File: sunkit_image/trace.py [hunk 25]
```diff
@@ -0,0 +1,195 @@
...
+image.
+"""
+
+import matplotlib.pyplot as plt
```
Comment (nabobalis) [inline sunkit_image/trace.py]: Is this used?

File: sunkit_image/trace.py [hunk 26]
```diff
@@ -0,0 +1,195 @@
...
+import numpy as np
+from scipy import interpolate
```
Comment (nabobalis) [inline sunkit_image/trace.py]:
```suggestion
from scipy import interpolate
```
Is this used?

File: sunkit_image/trace.py [hunk 27]
```diff
@@ -0,0 +1,195 @@
...
+
+from sunkit_image.utils import bandpass_filter, erase_loop_in_residual, curvature_radius, initial_direction_finding, loop_add
```
Comment (nabobalis) [inline sunkit_image/trace.py]:
```suggestion
from sunkit_image.utils import (bandpass_filter, erase_loop_in_residual, curvature_radius,
                                                initial_direction_finding, loop_add)
```

File: sunkit_image/trace.py [hunk 28]
```diff
@@ -0,0 +1,195 @@
...
+    lmin : `int`
+        The length of the smallest loop to be detected in pixels.
+    nstruc : `int`
+        Maximum limit of traced structures.
+    nloop : `int`
+        Maximum number of detected loops.
+    ngap : `int`
+        Number of pixels in the loop below the flux threshold.
+    qthresh1 : `float`
+        The ratio of image base flux and median flux. All the pixels in the image below `qthresh1 * median` intensity value
```
Comment (nabobalis) [inline sunkit_image/trace.py]: line is a bit too long

File: sunkit_image/trace.py [hunk 29]
```diff
@@ -0,0 +1,195 @@
...
+        are made to zero before tracing the loops.
+    qthresh2 : `float`
+        The factor which determines noise in the image. All the intensity values between `qthresh2 * median` are considered
```
Comment (nabobalis) [inline sunkit_image/trace.py]: line is a bit too long

File: sunkit_image/trace.py [hunk 30]
```diff
@@ -0,0 +1,195 @@
...
+        to be noise. The median for noise is chosen after the base level is fixed.
+    file : `bool`
+        If set to `True` an IDL style output txt file is created with the name as `loop.txt`
```
Comment (nabobalis) [inline sunkit_image/trace.py]:
```suggestion
        If set to `True` an IDL style output txt file is created with the name as ``loop.txt``.
```

File: sunkit_image/trace.py [hunk 31]
```diff
@@ -0,0 +1,195 @@
...
+
+    Returns
+    -------
+    `list`
+        A list of all loop where each element is itself a list of points containg ``x`` and ``y`` coordinates for each point.
```
Comment (nabobalis) [inline sunkit_image/trace.py]: line is a bit too long

File: sunkit_image/utils.py [hunk 32]
```diff
@@ -0,0 +1,517 @@
...
+    width : `int`
+        Width of the boxcar. The `width` should always be odd but if even value is given `width + 1` as the width of the boxcar.
+    nanopt : {"propagate" | "replace"}
+        It decides whether to `propagate` NAN's or `replace` them.
+
+    Returns
+    -------
+    `numpy.ndarray`
+        Smoothed image.
+    """
```
Comment (nabobalis) [inline sunkit_image/utils.py]: Should add the stackoverflow link as a reference here

File: sunkit_image/utils.py [hunk 33]
```diff
@@ -0,0 +1,517 @@
...
+def smooth(image, width, nanopt="replace"): 
+    """
+    Python implementation of the IDL `smooth <https://www.harrisgeospatial.com/docs/smooth.html>`__.
+
+    Parameters
+    ----------
+    image : `numpy.ndarray`
+        Image to be filtered.
+    width : `int`
+        Width of the boxcar. The `width` should always be odd but if even value is given `width + 1` as the width of the boxcar.
```
Comment (nabobalis) [inline sunkit_image/utils.py]: line is a bit too long

File: examples/tracing_loops.py (hunk 21 above)
Comment (nabobalis) [inline examples/tracing_loops.py]: There are no loops on this image.

Comment (vatch123) [reply examples/tracing_loops.py]: Yes we plot the image first on this the loops are overlayed. See the lines below

File: docs/code_ref/noise.rst (hunk 7 above)
Comment (vatch123) [reply docs/code_ref/noise.rst]: No idea I did not do anything to the noise module

File: examples/tracing_loops.py (hunk 17 above)
Comment (vatch123) [reply examples/tracing_loops.py]: I am not sure this was the best one which I found, if there is something better we can change it

File: examples/tracing_loops.py (hunk 20 above)
Comment (vatch123) [reply examples/tracing_loops.py]: How?? the return list is made in this particular way

File: examples/tracing_loops.py (hunk 21 above)
Comment (vatch123) [reply examples/tracing_loops.py]: I will write it in a better way

File: sunkit_image/trace.py (hunk 29 above)
Comment (vatch123) [reply sunkit_image/trace.py]: But it serves its purpose. I am unable to think of anything that I should remove

File: sunkit_image/utils.py (hunk 33 above)
Comment (vatch123) [reply sunkit_image/utils.py]: same here, what should I remove from here

Comment (nabobalis) [reply sunkit_image/utils.py]: You should just linebreak it into two lines.

File: examples/tracing_loops.py (hunk 17 above)
Comment (nabobalis) [reply examples/tracing_loops.py]: http://docs.sunpy.org/en/stable/code_ref/cm.html

I think its the top colour map here.

File: sunkit_image/trace.py (hunk 29 above)
Comment (nabobalis) [reply sunkit_image/trace.py]: Again, just linebreak it into two lines.

File: examples/tracing_loops.py [hunk 34]
```diff
@@ -0,0 +1,62 @@
...
+###########################################################################
+# Now the loop tracing will begin. We will be using the same set of parameters
+# as in the IDL tutorial.
+# The lowpass filter boxcar filter size ``nsm1`` is taken to be 3.
+# The minimum radius of curvature at any point in the loop ``rmin`` is 30 pixels.
+# The length of the smallest loop to be detected ``lmin`` is 25 pixels.
+# The maximum number of structures to be examined ``nstruc`` is 1000.
+# The number of extra points in the loop below noise level to terminate a loop tracing ``ngap`` is 0.
+# The base flux and median flux ratio ``qthresh1`` is 0.0
+# The noise threshold in the image with repect to median flux ``qthresh2`` is 3.0
```
Comment (nabobalis) [inline examples/tracing_loops.py]:
```suggestion
# The noise threshold in the image with repect to median flux ``qthresh2`` is 3.0 .
```

File: examples/tracing_loops.py [hunk 35]
```diff
@@ -0,0 +1,62 @@
...
+
+###########################################################################
+# Now the loop tracing will begin. We will be using the same set of parameters
+# as in the IDL tutorial.
+# The lowpass filter boxcar filter size ``nsm1`` is taken to be 3.
+# The minimum radius of curvature at any point in the loop ``rmin`` is 30 pixels.
+# The length of the smallest loop to be detected ``lmin`` is 25 pixels.
+# The maximum number of structures to be examined ``nstruc`` is 1000.
+# The number of extra points in the loop below noise level to terminate a loop tracing ``ngap`` is 0.
+# The base flux and median flux ratio ``qthresh1`` is 0.0
```
Comment (nabobalis) [inline examples/tracing_loops.py]:
```suggestion
# The base flux and median flux ratio ``qthresh1`` is 0.0.
```

File: examples/tracing_loops.py [hunk 36]
```diff
@@ -0,0 +1,62 @@
...
+# The noise threshold in the image with repect to median flux ``qthresh2`` is 3.0 
+loops = trace.occult2(image, nsm1=3, rmin=30, lmin=25, nstruc=1000, ngap=0, qthresh1=0.0, qthresh2=3.0, file=True)
```
Comment (nabobalis) [inline examples/tracing_loops.py]:
```suggestion
loops = trace.occult2(image, nsm1=3, rmin=30, lmin=25, nstruc=1000, 
                                  ngap=0, qthresh1=0.0, qthresh2=3.0, file=True)
```

File: sunkit_image/utils/utils.py [hunk 37]
```diff
@@ -101,10 +109,11 @@ def find_pixel_radii(smap, scale=None):
...
         entry in the input map data.
     """
-    # Calculate the co-ordinates of every pixel.
-    coords = all_coordinates_from_map(smap)
+    # Calculate all the x and y coordinates of every pixel in the map.
+    x, y = np.meshgrid(*[np.arange(v.value) for v in smap.dimensions]) * u.pix
 
-    # TODO: check that the returned coordinates are indeed helioprojective cartesian
+    # Calculate the helioprojective Cartesian co-ordinates of every pixel.
+    coords = smap.pixel_to_world(x, y).transform_to(frames.Helioprojective)
```
Comment (nabobalis) [inline sunkit_image/utils/utils.py]: Are these changes correct?

File: sunkit_image/utils/utils.py [hunk 38]
```diff
@@ -125,8 +134,8 @@ def get_radial_intensity_summary(
 
     Parameters
     ----------
-    smap : `~sunpy.map.Map`
-        A SunPy map.
+    smap : sunpy.map.Map
+        A sunpy map.
 
     radial_bin_edges : `~astropy.units.Quantity`
         A two-dimensional array of bin edges of shape (2, nbins) where nbins is
```
Comment (nabobalis) [inline sunkit_image/utils/utils.py]: Are these changes correct?

File: sunkit_image/utils/utils.py [hunk 39]
```diff
@@ -173,3 +182,439 @@ def get_radial_intensity_summary(
...
+    # al[ip + 1] = al[ip] + sign_dir * (step / rad_i[v])
+    # ir[ip+1] = ib1 + v
+
+    # al_mid = (al[ip]+al[ip+1]) / 2.
+
+    # xl[ip+1] = xl[ip] + step * np.float32(np.cos(al_mid + np.pi * idir))
+    # yl[ip+1] = yl[ip] + step * np.float32(np.sin(al_mid + np.pi * idir))
+    # ix_ip = min(max(int(xl[ip + 1] + 0.5), 0), nx - 1)
+    # iy_ip = min(max(int(yl[ip + 1] + 0.5), 0), ny - 1)
+    # zl[ip + 1] = image[ix_ip, iy_ip]
```
Comment (nabobalis) [inline sunkit_image/utils/utils.py]: What is all this?

Comment (vatch123) [reply sunkit_image/utils/utils.py]: This is just an attempt to vectorize the codes to remove the loop such that the code speeds up. If you look the `initial_direction_finding` function above this and compare it across this commit you will see I have removed the loop in it which decreased the example run time by almost 20secs. I am trying to do that for this also but this is tricky. I will try to fix it if possible otherwise I will just remove this block.

File: examples/tracing_loops.py [hunk 40]
```diff
@@ -0,0 +1,63 @@
+"""
+=====================
+Coronal Loops Tracing
+=====================
+
+This example applies traces out the coronal loops in a FITS file
```
Comment (wafels) [inline examples/tracing_loops.py]: "This example traces out coronal loops in a FITS image"

File: examples/tracing_loops.py [hunk 41]
```diff
@@ -0,0 +1,63 @@
...
+using `~sunkit_image.trace.occult2`. In this example we will use the settings
+and the data from Markus Aschwanden's tutorial on his IDL implementation of
+the ``OCCULT2`` algorithm, which can be found `here <http://www.lmsal.com/~aschwand/software/tracing/tracing_tutorial1.html>`__.
+The aim of this example is to demonstrate that `~sunkit_image.trace.occult2` provides similar
+results as the IDL version.
```
Comment (wafels) [inline examples/tracing_loops.py]: 'results compared to the IDL version."

File: examples/tracing_loops.py [hunk 42]
```diff
@@ -0,0 +1,63 @@
...
+import numpy as np
+
+import sunkit_image.trace as trace
+import sunpy.cm
+
+###########################################################################
+# We will be using `astropy.io.fits.getdata` to read the FITS file from the tutorial website.
+image = astropy.io.fits.getdata("http://www.lmsal.com/~aschwand/software/tracing/TRACE_19980519.fits", ignore_missing_end=True)
+
+# The original image is plotted to showcase the difference.
```
Comment (wafels) [inline examples/tracing_loops.py]: "The original image is plotted, showing coronal loops."

File: examples/tracing_loops.py [hunk 43]
```diff
@@ -0,0 +1,63 @@
...
+plt.imshow(image, cmap="hmimag", origin="lower")
+
+###########################################################################
+# Now the loop tracing will begin. We will be using the same set of parameters
```
Comment (wafels) [inline examples/tracing_loops.py]: "We will use the same set of parameters"

File: examples/tracing_loops.py [hunk 44]
```diff
@@ -0,0 +1,63 @@
...
+# as in the IDL tutorial.
+# The lowpass filter boxcar filter size ``nsm1`` is taken to be 3.
+# The minimum radius of curvature at any point in the loop ``rmin`` is 30 pixels.
+# The length of the smallest loop to be detected ``lmin`` is 25 pixels.
+# The maximum number of structures to be examined ``nstruc`` is 1000.
+# The number of extra points in the loop below noise level to terminate a loop tracing ``ngap`` is 0.
+# The base flux and median flux ratio ``qthresh1`` is 0.0.
+# The noise threshold in the image with repect to median flux ``qthresh2`` is 3.0 .
+loops = trace.occult2(image, nsm1=3, rmin=30, lmin=25, nstruc=1000,
```
Comment (wafels) [inline examples/tracing_loops.py]: 'For the meaning of these parameters please consult the OCCULT2 article"

File: examples/tracing_loops.py [hunk 45]
```diff
@@ -0,0 +1,63 @@
...
+                      ngap=0, qthresh1=0.0, qthresh2=3.0, file=True)
+
+###############################################################################
+# `~sunkit_image.trace.occult2` returns a list of all loop where each element
+# is itself a list of points containg ``x`` and ``y`` coordinates for each point.
```
Comment (wafels) [inline examples/tracing_loops.py]: containg -> containing

File: sunkit_image/tests/test_trace.py [hunk 46]
```diff
@@ -0,0 +1,155 @@
...
+    # Testing on the same input files as in the IDL tutorial
+    loops = trace.occult2(image, nsm1=3, rmin=30, lmin=25, nstruc=1000, ngap=0, qthresh1=0.0, qthresh2=3.0, file=True)
+
+    result = np.loadtxt("loops.txt")
+
+    os.remove("loops.txt")
+
+    # First we will verify on the Python version of the output which we know is correct as we had plotted
+    # it earlirer and also have made visual comparisons
+    expect = np.loadtxt(filepath_py)
```
Comment (nabobalis) [inline sunkit_image/tests/test_trace.py]: Use https://pypi.org/project/pytest-arraydiff/ to do these tests instead of opening the file.

File: sunkit_image/trace.py [hunk 47]
```diff
@@ -0,0 +1,218 @@
...
+from sunkit_image.utils import (bandpass_filter, erase_loop_in_image, curvature_radius,
+                                initial_direction_finding, loop_add)
+
+__all__ = ["occult2"]
+
+
+def occult2(image, nsm1, rmin, lmin, nstruc, ngap, qthresh1, qthresh2, file=False):
+    """
+    Implements the Oriented Coronal CUrved Loop Tracing (OCCULT-2) algorithm
+    for loop tracing in solar images.
```
Comment (nabobalis) [inline sunkit_image/trace.py]:
```suggestion
    for loop tracing in images.
```

**Comment by vatch123** [comment]
Changes from wafels have been incorporated

File: sunkit_image/utils/utils.py [hunk 48]
```diff
@@ -173,3 +181,395 @@ def get_radial_intensity_summary(
             for i in range(0, nbins)
         ]
     )
+
+
+# The functions below this are subroutines for the OCCULT 2.
```
[the remark this answers is not in the document]
Comment (vatch123) [reply sunkit_image/utils/utils.py]: I can do that if @nabobalis agrees to it.

**Comment by nabobalis** [comment]
pep8speaks is complaining, any chance you can look into them and fix it?

**Comment by wafels** [comment]
Approved.

**Comment by nabobalis** [comment]
Thanks everyone for the reviews and @vatch123 for the PR!
